### DISCUSSION TITLE
Select on an RPC 'view' and get a foreign key relation

### DISCUSSION THREAD

**Discussion by tourbillonlabs** [supabase/supabase#9091] [Questions] [answered]
Hello!

This is more of a sounding board query, so I welcome your thoughts.

## Context

I have the following structure (bare minimum fields to illustrate).

```
tableA (
  id uuid
  b_id int8 NOT NULL REFERENCES tableB (id)
  user_id uuid 
)

tableB (
  id bigserial
  c_id int8 NOT NULL REFERENCES tableC (id)
)

tableC (
  id bigserial
)
```

I effectively want to do this: `SELECT *, [tableC relation] FROM tableB b WHERE b.id IN (SELECT DISTINCT a.b_id FROM tableA a WHERE a.user_id = user_id)`.

With the select relation syntax offered by supabase/postgrest I can functionally achieve this with two supabase queries:

```
// SELECT DISTINCT a.b_id FROM tableA a WHERE a.user_id = user_id
const { data: ids } = await supabaseClient
    .rpc("distinct_b_ids", { user_id })
    .select("id");

// SELECT *, [tableC relation] FROM tableB b WHERE b.id IN ([the above SELECT DISTINCT])
await supabaseClient
    .from("tableB")
    .select("*, tableC:c_id(*)")
    .in("id", ids)
```

Note: Based on other posts, I have established that I cannot `SELECT DISTINCT` using this lib which is why I have used an RPC to fetch the ids for the second query.

## Query

*It would be nice* to be able to do this in one supabase call instead.

Searching supabase and postgrest suggests I should create a RPC (not VIEW because I need the `user_id` param) so that I can do the `SELECT * FROM tableB b WHERE b.id IN (SELECT DISTINCT a.b_id FROM tableA a WHERE a.user_id = user_id)` part.

`RETURN TABLE` using the RPC doesn't allow me to associate a foreign key, and if I try something like:

```
await supabaseClient
    .rpc("user_tableB_view", { user_id })
    .select("*, tableC:c_id(*)")
```

I do in fact get the error `"Could not find a relationship between 'record' and 'c_id' in the schema cache"`. I gather it's because you can't dynamically assign fk relationships in this way, which is reasonable. Bear in mind I also tried creating a FK in the `RETURN TABLE` definition and get a syntax error.

Am I correct in my assumption that I cannot use the relation select syntax in this way and would have to do an old fashioned `JOIN` if I wanted to do this in a single query?

But, I also see this post about postgrest v10 (https://supabase.com/blog/postgrest-v10) and computed relationships (https://postgrest.org/en/v10.0/api.html#computed-relationships) and I'm wondering if this would allow me to query relations on an RPC/VIEW. 

Thanks!

**Comment by steve-chavez** [comment, marked as answer]
>  RETURN TABLE definition

Would it be possible to return the table type you want to join on instead?(`RETURNS SETOF tableB`) Then embedding should work out of the box as mentioned in https://postgrest.org/en/stable/api.html#embedding-on-stored-procedures

**Reply by tourbillonlabs** [reply]
I have overthought this.

You are a star. Thank you.
